File: app/src/main/java/com/example/myapplication/data/ProDetailsRepository.java
package com.example.myapplication.data;

import com.example.myapplication.data.bean.ProDetails;
import com.example.myapplication.data.network.ProDetailsNetwork;

public class ProDetailsRepository {
    private ProDetailsNetwork network;

    public ProDetailsRepository(ProDetailsNetwork network) {
        this.network = network;
    }

    public ProDetails getProDetails(String category) {
        return network.fetchProDetails(category);
    }
}

File: app/src/main/java/com/example/myapplication/data/network/ProDetailsRepository.kt
package com.example.myapplication.data.network

import com.example.myapplication.data.bean.ProDetailsInfo

class ProDetailsRepository(private val network: ProDetailsNetwork) {
    suspend fun getProDetails(id: Int): ProDetailsInfo {
        return network.fetchProDetails(id)
    }

    companion object {
        private lateinit var instance: ProDetailsRepository
        fun getInstance(network: ProDetailsNetwork): ProDetailsRepository {
            //可以通过属性引用的 isInitialized来检测该lateinit var是否已初始化   this:: 用来访问类的成员，this可以省略
            if (!::instance.isInitialized) {
                synchronized(ProDetailsRepository::class.java) {
                    if (!::instance.isInitialized) {
                        instance = ProDetailsRepository(network)
                    }
                }
            }
            return instance
        }
    }
}